### ISSUE TITLE
Add dependency locking setup for Android

### ISSUE THREAD

**Issue by halaei** [react-native-community/discussions-and-proposals#443]
### Introduction
The current React Native initial setup doesn't provide any dependency locking for Android. To me it is a very hard task to setup a correct mechanism for locking dependencies on Android, not only because it is not as easy as how `yarn.lock` file is generated, but also because as a React Native developer I am not really supposed to be an expert Java/Android developer, so I am not familiar with the tools. Therefore, an official documentation on how to achieve a reasonable dependency locking mechanism for React Native projects is required. Even better, we can have the dependency locking already set up in the next versions of React Native, by default. 

### Details

My motivation is simple. I see a bug in the current version of our app. I want to checkout an older version to see how the app used to work and compare the behavior. But the older version doesn't compile because android dependencies were not locked and some unknown library released a breaking update and I don't know 
1. The exact library that causes the problem.
2. What version is compatible with my app.
3. How to use the compatible version of the dependency.

### Discussion points
We can use either of the followings. But I can't figure out the exact configuration needed for a React Native project:

#### 1. https://docs.gradle.org/current/userguide/project_report_plugin.html
Setup:
In `app.build.gradle`
```
apply plugin: 'project-report'
```
`cd android && ./gradlew projectReport` will produce a report in `/android/app/build/reports/project/` directory. The report includes a dependencies.txt file which contains all the Android dependencies and their versions. We can move the file one directory up and commit in into our version control repository. However, this is not a lockfile. It's just a report.
#### 2. https://docs.gradle.org/current/userguide/dependency_locking.html
I don't know how to set it up to lock all the dependencies. Too complicated for me.
#### 3. https://github.com/nebula-plugins/gradle-dependency-lock-plugin
Setup:
In `android/build.gradle`
```
buildscript {
    ...
    dependencies {
        ...
        classpath "com.netflix.nebula:gradle-dependency-lock-plugin:11.3.2" // for gradle 7, switch to 12.+
    }
}
```
In `app/build.gradle`
```
apply plugin: "nebula.dependency-lock"
```
Running `cd android && ./gradlew generateLock saveLock`generates a json lockfile that we can commit. However, I find out that the lock is not complete. Not all the dependencies reported by the 'project-report' plugin are listed in the output lock. If someone knows how to fix my setup, please let me know.

### iOS
The iOS setup uses pod, which is perfect regarding dependency locking.

**Comment by cortinico** [comment]
Thanks for the details issue @halaei 

> Some gradle dependency releases a breaking update and suddenly my React Native app stops to be compiled.

Did this really happened to you? If yes, which dependency caused this issue?
I'm asking as the majority of Gradle dependencies we specify have a fixed version. What can happen is a transitive dependency pulling in a newer version of a library.

As a rule of thumb, I would discourage using dynamic versions in your `build.gradle` file (i.e. replace the `+` versions with a specific version).

As you already pointed out, Gradle already supports having a lockfile. I suggest to follow the guide here: https://docs.gradle.org/current/userguide/dependency_locking.html

I agree that the Gradle documentation over there could be a bit complicated. I can suggest to [checkout this section](https://youtu.be/GtZ-WZWoneQ?t=1198) of a talk I gave where I mention how Gradle Dependency Locking works and also give some examples. I hope you find it useful.

Finally, I don't think this should be enabled by default on all the RN apps. Dependency locking, specifically on Android, tend to be a niche feature. Most of the times you don't really need it and it would create unnecessary complexity for our users.

**Comment by halaei** [comment]
@cortinico 

> Did this really happened to you?

Yes. It was play-services-ads, that was obviously required by some third party packages, using + versions as you said. A similar issue may also happen with updating package.lock/yarn.lock. If I have 2 package.json dependencies that both import a gradle dependency and one of them updates the version, the other package may break.

Thanks for the docs. Your video looks like a great one. I will definitely watch it soon. But I wasn't able to make a locking mechanism that is as thorough as the report generated by `project-report`. I guess it is a little bit harder than it should be for sub-projects and the normal android or react native projects.

**Comment by ejain** [comment]
Just happened to us when one morning `androidx.core:core` started resolving to `1.12.0-alpha05` instead of `1.10.0`. Might not have noticed if our builds hadn't failed because the newer version requires a higher `compileSdkVersion`...

Note that we don't have a direct dependency on that package; I think it gets pulled in with `android.useAndroidX=true`.

**Comment by cortinico** [comment]
> Note that we don't have a direct dependency on that package; I think it gets pulled in with `android.useAndroidX=true`.

Nope that's not the case. You definitely have one or more module in your project that are depending on it. You can inspect who's depending on it by running a Gradle Scan (run a build command with `--scan`).

**Comment by ejain** [comment]
@cortinico I did just that, but no module is willing to accept responsibility for pulling in the `1.12.0-alpha05` version 🙃

<img width="572" alt="dependencies" src="https://github.com/react-native-community/discussions-and-proposals/assets/1323903/adbbbd37-ca3d-4f1b-bd50-5e41c36e2f86">

**Comment by ejain** [comment]
@liamjones I did that, too, and `androidx.core:core` appears as a top-level dependency...

**Comment by cortinico** [comment]
> @liamjones I did that, too, and `androidx.core:core` appears as a top-level dependency...

Can you share your build scan?

**Comment by cortinico** [comment]
Thanks @ejain 

<img width="1030" alt="Screenshot 2023-06-13 at 19 32 26" src="https://github.com/react-native-community/discussions-and-proposals/assets/3001957/3f2a8af6-d03f-459e-a38f-f1010187d6f2">

The problem is that you have in your app module, a dependency on `androidx.appcompat:appcompat:+` which is causing this transitive dependency to be pulled in.

React Native depends on `androidx.appcompat:appcompat` but on a specific version (`1.4.1` to be precise). Your project is overriding it as you have a `+` somewhere in your android/app/gradle file (or you have a plugin that is applying it).
